### ISSUE TITLE
WhatsApp Web stopped working around 2018-11-30 - Asks for Google Chrome ≥ 36

### ISSUE THREAD

**Issue by prashant-shahi** [nativefier/nativefier#719]
### Description

WhatsApp Web was working so well until few days back. Now when the application is started, it says the following:
```
WhatsApp works with Google Chrome 36+
To use WhatsApp, update Chrome or use Mozilla Firefox, Safari, Microsoft Edge or Opera.

Update Google Chrome
```

![Natifier WhatsApp Web](http://imgur.com/n39ofyrl.png)

### Steps to reproduce issue

$ `nativefier --name "WhatsApp Web" https://web.whatsapp.com`
$ `./whatsapp-web-linux-x64/whatsapp-web`

### Details

- Are you nativefying a *public* website? **Yes**.
- Feature request? **No**.
- Full nativefier command used to build your app: `nativefier --name "WhatsApp Web" https://web.whatsapp.com`
- Version of Nativefier (run `nativefier --version`): `v7.6.8`
- Version of node.js (run `node --version`): `v8.10.0`
- OS: `Ubuntu 18.04.1 LTS - Bionic`
- Error message / stack trace (if any): **None**.

**Comment by ronjouch** [comment]
@coolboi567 just re-package the app. Yes we know we should do a better job at communicating this to users, see https://github.com/jiahaog/nativefier/issues/556; contributions welcome.

**Comment by bitlab-code** [comment]
same issue with vers. 7.6.8

**Comment by ronjouch** [comment]
@bitlab-code Nativefier 7.6.8 ships with a default Electron version of 3.0.3, based on Chrome 66. Are you sure you did re-package your app with it, by re-running the nativefier command used to package it? If yes, then inside your nativefied app can you pop open the inspector (<kbd>Ctrl/Cmd</kbd>+<kbd>Shift</kbd>+<kbd>I</kbd>) and in the console type `navigator.userAgent` . What does it output?

**Comment by bitlab-code** [comment]
@ronjouch command + log
<img width="1238" alt="2018-11-30 02 56 24 pm" src="https://user-images.githubusercontent.com/19858243/49293340-3119dc80-f4b0-11e8-8884-8341227f7ed6.png">

**Comment by ronjouch** [comment]
@bitlab-code thanks, your setup looks good indeed.

WhatsApp must be doing some kind of new userAgent sniffing / feature detection. Patches welcome.

**Comment by lima-eduardo** [comment]
As a "workaround" you can go to 'edit > clear app data' and rescan the QR code.

**Comment by saulocastelo** [comment]
> As a "workaround" you can go to 'edit > clear app data' and rescan the QR code.

**Better "workaround"**

Steps:
* Open DevTools (Ctrl+Shift+i)
* Application
* Service Workers 
* Bypass for network
* Reload (Ctrl+r)
* Close DevTools

Or inject this code
```javascript
if (document.body.innerText.replace(/\n/g, ' ').search(/whatsapp works with.*to use whatsapp.*update/i) === 0)
	navigator.serviceWorker.getRegistration().then(function (r) { r.unregister(); document.location.reload() });
```

**Comment by vip9** [comment]
"Bypass for network" works for me only once.

**Comment by shvchk** [comment]
@lipe1966 create file `whatsapp-fix.js` with inject code from @saulocastelo comment, then `nativefier -n "WhatsApp" --inject whatsapp-fix.js "https://web.whatsapp.com"`

**Comment by roccomuso** [comment]
I ran it clean using the container to build a new bin and worked. So what's going on exactly?
I had that error with Chrome 66. (does it apply only with old existing sessions?)

**Comment by Birkenstab** [comment]
It does not work for me, even with a new session. But if I open multiple instances of the app at the same time, one usually works.
With the fix saulocastelo it works all the time, as far as I can tell right now. Thanks @saulocastelo :)

**Comment by roccomuso** [comment]
I'm having the same issue on electron, but that @saulocastelo 's fix doesn't work. in the promise result `r` is undefined for me.

**Comment by eaorak** [comment]
> As a "workaround" you can go to 'edit > clear app data' and rescan the QR code.

This only works once. @shvchk's solution worked for me.

**Comment by roccomuso** [comment]
Related: https://github.com/meetfranz/franz/issues/1185
Hopefully someone will find a stable fix

**Comment by lipe1966** [comment]
Found this snap that works perfectly https://snapcraft.io/whatsdesk

**Comment by FcoMarcosMAbreu** [comment]
> @lipe1966 create file `whatsapp-fix.js` with inject code from @saulocastelo comment, then `nativefier -n "WhatsApp" --inject whatsapp-fix.js "https://web.whatsapp.com"`

Thank you very much! This works for me too.

**Comment by saulocastelo** [comment]
> I'm having the same issue on electron, but that @saulocastelo 's fix doesn't work. in the promise result `r` is undefined for me.

Are you still facing this problem? Maybe I can help you, send me a message using Google Hangout or an E-mail with the details (saulocastelo@gmail.com)

**Comment by saulocastelo** [comment]
> > As a "workaround" you can go to 'edit > clear app data' and rescan the QR code.
> 
> This only works once. @shvchk's solution worked for me.

You have to activate on every session, that's the reason that i make the js code

**Comment by saulocastelo** [comment]
> Hello! I hope I can contribute with my suggestion. You could enable the right click and show the element inspector option, among others. Through the element inspector, the user can define a user agent supported by WhatsApp Web. I am looking forward to using Nativefier. Success to all!

The nativefier has an option('-u')  to set the useragent but didn't work

**Comment by has7eem** [comment]
I'm having this problem now (Electron), did you find her solution?

**Comment by bitlab-code** [comment]
> I'm having this problem now (Electron), did you find her solution?

Occasionally it happens, When it happens I always resolve with this:
> As a "workaround" you can go to 'edit > clear app data' and rescan the QR code.

**Comment by has7eem** [comment]
@bitlab-code  
Sorry for not explaining more
I'm trying to open WhatsApp in my electron app.

**Comment by has7eem** [comment]
I have tried many solutions, and that's just who worked with me.

```html
<webview 
src="https://web.whatsapp.com/" 
partition="persist:github" 
useragent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36">
</webview>
```
Do not forget to delete your storage first.
If you don't care about save your registration and don't need to clear storage every time.
read about `partition` in Electron Doc

**Comment by Jodanzuki** [comment]
@saulocastelo 's solution worked for quite some time, but as of yesterday, only Clearing App Data has been doing the trick. But I don't want to rescan the code everytime.

**Comment by roccomuso** [comment]
@Jodanzuki agree, stopped working again yesterday. Clean app required now. too bad.

**Comment by saulocastelo** [comment]
> @saulocastelo 's solution worked for quite some time, but as of yesterday, only Clearing App Data has been doing the trick. But I don't want to rescan the code everytime.

I will see this problem again soon as possible and I will get back to you!

**Comment by saulocastelo** [comment]
> @Jodanzuki agree, stopped working again yesterday. Clean app required now. too bad.

Please always add me to the conversation when this has any failure, will be easier for me to see your problem. 😉🖖🏻

**Comment by boban-dj** [comment]
Same here:
when using your whats-app-linux-64 it's possible to enter dev tools on chromium: ctrl+shift+I, then in Applications -> Bypass for network check, then it works. But need to do it everytime when start.
Do you have solution?

**Comment by has7eem** [comment]
I have tried many solutions, and that's just who worked with me.

```html
<webview 
src="https://web.whatsapp.com/" 
useragent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36">
</webview>
```
Do not forget to delete your storage first.
If you don't care about save your registration and don't need to clear storage every time.
read about `partition` in Electron Doc

**Comment by saulocastelo** [comment]
Heeeey guys are whaaaaat are you waiting foooooor 

`if (document.body.innerText.replace(/\n/g, ' ').search(/whatsapp works with.*to use whatsapp.*update/i) !== -1)
	navigator.serviceWorker.getRegistration().then(function (r) { r.unregister(); document.location.reload() });`


just replace === 0 to !== -1 and be happy!

**Comment by saulocastelo** [comment]
> Hi everyone!
> 
> I decided to create "fix" for WhatsApp which will build the application correctly. This is a simple bash script, which download JS script and execute build app with nativefier including JS file. I know is it not resolved bug, but at least the application works correctly.
> 
> https://github.com/DocBox12/WhatsApp-nativefier-fix

@DocBox12 Check the pull request!

**Comment by DocBox12** [comment]
@saulocastelo I accepted pull request. Thanks!

**Comment by fabionogueira** [comment]
This is work for me:
```javascript
mainWindow.webContents.on('dom-ready', function(e) {
        let script = `
            window.onunload = () => {
                
                navigator.serviceWorker.getRegistrations().then(
                    function(registrations) {
                        for(let registration of registrations) {
                            registration.unregister();
                        }
                    }
                )

            }
            `
        mainWindow.webContents.executeJavaScript(script)
    })
```

**Comment by saulocastelo** [comment]
> This is work for me:
> 
> ```js
> mainWindow.webContents.on('dom-ready', function(e) {
>         let script = `
>             window.onunload = () => {
>                 
>                 navigator.serviceWorker.getRegistrations().then(
>                     function(registrations) {
>                         for(let registration of registrations) {
>                             registration.unregister();
>                         }
>                     }
>                 )
> 
>             }
>             `
>         mainWindow.webContents.executeJavaScript(script)
>     })
> ```

This is a great way to do but will always remove the workers even if there's no need. But will avoid the reload on startup. Another issue could be if you need to kill the app (unresponsive) ... the window.unload will not be called and then you will have the problem again.

Tell me if you are able to solve this 2 questions. It's really better avoid the reload when start.

Thank's for sharing this. :)

**Comment by chriskyfung** [comment]
Did WhatsApp update again?
It keeps asking for Google Chrome 36+, no matter I refresh the page manually or programmingly.

**Comment by raulcraveiro** [comment]
Now It's asking for me again, to update for Chrome 49+. Tried a lot of fixes that I saw here but nothing works. :(

**Comment by MauricioGlz** [comment]
I have to second this, the issue is back, whatsapp is asking for something already has. No big deal having to do the developer tools stuff, but it's frustrating to do it every single time I open the app.

> Now It's asking for me again, to update for Chrome 49+. Tried a lot of fixes that I saw here but nothing works. :(

**Comment by filipegarrett** [comment]
I can confirm also that the issue is back. Nativefier is up to date here.

**Comment by muj-programmer** [comment]
> Now It's asking for me again, to update for Chrome 49+. Tried a lot of fixes that I saw here but nothing works. :(

Facing the same issue here

**Comment by CrazyEasy** [comment]
I had the same issue again, since 2 weeks or so.

> Hi everyone!
> 
> I decided to create "fix" for WhatsApp which will build the application correctly. This is a simple bash script, which download JS script and execute build app with nativefier including JS file. I know is it not resolved bug, but at least the application works correctly.
> 
> https://github.com/DocBox12/WhatsApp-nativefier-fix

This fix actually DID still work for me!

**Comment by filipegarrett** [comment]
> I had the same issue again, since 2 weeks or so.
> 
> > Hi everyone!
> > I decided to create "fix" for WhatsApp which will build the application correctly. This is a simple bash script, which download JS script and execute build app with nativefier including JS file. I know is it not resolved bug, but at least the application works correctly.
> > https://github.com/DocBox12/WhatsApp-nativefier-fix
> 
> This fix actually DID still work for me!

Worked here as well!

**Comment by muj-programmer** [comment]
> > Now It's asking for me again, to update for Chrome 49+. Tried a lot of fixes that I saw here but nothing works. :(
> 
> Facing the same issue here

updated the electron version to 8.0.0 and its working fine for me

**Comment by ronjouch** [comment]
I just released Nativefier 8.0.2, which defaults to latest Electron 8.1.1. I don't use WhatsApp, can anyone test it and report if this issue is addressed, and if the fix at https://github.com/jiahaog/nativefier/issues/719#issuecomment-482736304 is still necessary.

**Comment by raulcraveiro** [comment]
> I just released Nativefier 8.0.2, which defaults to latest Electron 8.1.1. I don't use WhatsApp, can anyone test it and report if this issue is addressed, and if the fix at [#719 (comment)](https://github.com/jiahaog/nativefier/issues/719#issuecomment-482736304) is still necessary.

I'm trying to install nativefier here and getting an error:

`npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! nativefier@8.0.2 postinstall: `cd app && yarn install --no-lockfile --no-progress --silent`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the nativefier@8.0.2 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-03-15T22_51_32_660Z-debug.log`

**Comment by ronjouch** [comment]
> I'm trying to install nativefier here and getting an error:
> 
>
> ```
> npm ERR! code ELIFECYCLE npm ERR! errno 1 npm ERR! nativefier@8.0.2 postinstall: `cd app && yarn install --no-lockfile --no-progress --silent`
> npm ERR! Exit status 1
> npm ERR!
> npm ERR! Failed at the nativefier@8.0.2 postinstall script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR! /root/.npm/_logs/2020-03-15T22_51_32_660Z-debug.log`
> ```

**Comment by ronjouch** [comment]
> I'm trying to install nativefier here and getting an error:
> 
>
> ```bash
> npm ERR! code ELIFECYCLE npm ERR! errno 1
> npm ERR! nativefier@8.0.2 postinstall: `cd app && yarn install --no-lockfile --no-progress --silent`
> npm ERR! Exit status 1
> npm ERR!
> npm ERR! Failed at the nativefier@8.0.2 postinstall script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR! /root/.npm/_logs/2020-03-15T22_51_32_660Z-debug.log`
> ```

@raulcraveiro what does `/root/.npm/_logs/2020-03-15T22_51_32_660Z-debug.log` say? What happens if you manually do the `postinstall`: going in your global lib/node_modules/nativefier folder, then running `cd app && yarn install --no-lockfile --no-progress --silent`.

Installation works for me, but when installing locally in a user-writable path, not as root (which I then add to my PATH):

```bash
~ npm i -g nativefier
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
/home/ronj/.node_modules/bin/nativefier -> /home/ronj/.node_modules/lib/node_modules/nativefier/lib/cli.js

> core-js@3.6.4 postinstall /home/ronj/.node_modules/lib/node_modules/nativefier/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

> nativefier@8.0.2 postinstall /home/ronj/.node_modules/lib/node_modules/nativefier
> cd app && yarn install --no-lockfile --no-progress --silent

+ nativefier@8.0.2
added 288 packages from 332 contributors in 14.917s
```

**Comment by raulcraveiro** [comment]
This is the result of the debug.log: https://paste.ubuntu.com/p/Q5mZ28673x/

**Comment by ronjouch** [comment]
> This is the result of the debug.log: [paste.ubuntu.com/p/Q5mZ28673x](https://paste.ubuntu.com/p/Q5mZ28673x/)

@raulcraveiro thanks. So,

```
8682 verbose lifecycle nativefier@8.0.2~postinstall: unsafe-perm in lifecycle false
8683 verbose lifecycle nativefier@8.0.2~postinstall: PATH: /usr/share/npm/node_modules/npm-lifecycle/node-gyp-bin:/usr/local/lib/node_modules/nativefier/node_modules/.bin:/usr/local/lib/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
8684 verbose lifecycle nativefier@8.0.2~postinstall: CWD: /usr/local/lib/node_modules/nativefier
8685 silly lifecycle nativefier@8.0.2~postinstall: Args: [ '-c',
8685 silly lifecycle   'cd app && yarn install --no-lockfile --no-progress --silent' ]
8686 silly lifecycle nativefier@8.0.2~postinstall: Returned: code: 1  signal: null
8687 info lifecycle nativefier@8.0.2~postinstall: Failed to exec postinstall script
8688 verbose unlock done using /root/.npm/_locks/staging-3a08f0df5026584d.lock for /usr/local/lib/node_modules/.staging
8689 verbose stack Error: nativefier@8.0.2 postinstall: `cd app && yarn install --no-lockfile --no-progress --silent`
8689 verbose stack Exit status 1
8689 verbose stack     at EventEmitter.<anonymous> (/usr/share/npm/node_modules/npm-lifecycle/index.js:283:16)
8689 verbose stack     at EventEmitter.emit (events.js:189:13)
8689 verbose stack     at ChildProcess.<anonymous> (/usr/share/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
8689 verbose stack     at ChildProcess.emit (events.js:189:13)
8689 verbose stack     at maybeClose (internal/child_process.js:970:16)
8689 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
```

At 8684 we're told we're in `/usr/local/lib/node_modules/nativefier` (as root, right? You ran `sudo npm install -g nativefier`, correct?). Then `cd app && yarn install --no-lockfile --no-progress --silent` errors and returns `1`. Can you retry this manually?

```bash
cd /usr/local/lib/node_modules/nativefier/app
sudo yarn install --no-lockfile --no-progress --silent
# If `yarn` isn't installed globally, please run it wih its fully-qualified path,
# probably /usr/local/bin/yarn or node /usr/local/lib/node_modules/yarn/index.js
```

**Comment by raulcraveiro** [comment]
Yes, I ran with sudo.

When I try to cd to this path, It says that the directory doesn't exist D:

**Comment by ronjouch** [comment]
> When I try to cd to this path, It says that the directory doesn't exist D:

@raulcraveiro what's the output of `/usr/local/lib/node_modules/nativefier/` ?

**Comment by raulcraveiro** [comment]
`bash: /usr/local/lib/node_modules/nativefier/: Arquivo ou diretório inexistente`

The directory doesn't exist

**Comment by ronjouch** [comment]
@raulcraveiro sorry, I meant `ls /usr/local/lib/node_modules/nativefier/` . But wait, you say `/usr/local/lib/node_modules/nativefier/` itself doesn't exist?! This is weird, as your logs say this is where nativefier got installed:

```bash
8684 verbose lifecycle nativefier@8.0.2~postinstall: CWD: /usr/local/lib/node_modules/nativefier
```

Can you look around (use `find` or other filesystem search tools) and look where nativefier got installed?

**Comment by raulcraveiro** [comment]
@ronjouch The only file named nativefier is on /usr/bin, but there isn't any folder for nativefier.

**Comment by ronjouch** [comment]
> The only file named nativefier is on /usr/bin, but there isn't any folder for nativefier.

@raulcraveiro where does `/usr/bin/nativefier` point to? Pls `ls -la /usr/bin/nativefier`

**Comment by raulcraveiro** [comment]
@ronjouch lrwxrwxrwx 1 root root 41 mar 15 20:19 /usr/bin/nativefier -> ../lib/node_modules/nativefier/lib/cli.js

**Comment by ronjouch** [comment]
@raulcraveiro so, the link is broken now, right? Resolving the relative symlink gets us `/usr/lib/node_modules/nativefier/lib/cli.js`, and we said `/usr/lib/node_modules/nativefier` doesn't exist.

Can you try uninstalling nativefier globally, then retrying? `npm uninstall -g nativefier && npm install -g nativefier`

**Comment by raulcraveiro** [comment]
I uninstalled and installed again and I got the same error, and the /usr/local/lib/node_modules/nativefier/ still doesn't exist :/

**Comment by ronjouch** [comment]
Okay, someone else reported the same at https://github.com/jiahaog/nativefier/issues/923 . Stopping the discussion here to stop polluting the WhatsApp issue, and continuing the discussion at https://github.com/jiahaog/nativefier/issues/923

**Comment by raulcraveiro** [comment]
@ronjouch Just tried to create the Whatsapp nativefier with the 8.0.4 without the inject code. On the first time that I opened, it worked ok. But If I close the app and open again, the problem continues. :/

**Comment by ronjouch** [comment]
> @ronjouch Just tried to create the Whatsapp nativefier with the 8.0.4 without the inject code. On the first time that I opened, it worked ok. But If I close the app and open again, the problem continues. :/

@raulcraveiro hmmmmm, now we're back to a WhatsApp-specific issue, and here you're on your own, I don't use WhatsApp and don't wish to invest time debugging it. Maybe the serviceWorker-patching inject is still necessary. If not and if someone here finds out a generic change fixing the issue, patch welcome 🙂.

**Comment by alexkrass76** [comment]
I try to create Whatsapp using nativefier 8.0.4, the problem is the same, I used versions of Electron 8.1.1; 9.0.0-beta.9, used whatsapp-fix.js, no result, requires updating chrome

**Comment by alexkrass76** [comment]
> > As a "workaround" you can go to 'edit > clear app data' and rescan the QR code.
> 
> **Better "workaround"**
> 
> Steps:
> 
> * Open DevTools (Ctrl+Shift+i)
> * Application
> * Service Workers
> * Bypass for network
> * Reload (Ctrl+r)
> * Close DevTools
> 
> Or inject this code
> 
> ```js
> if (document.body.innerText.replace(/\n/g, ' ').search(/whatsapp works with.*to use whatsapp.*update/i) === 0)
> 	navigator.serviceWorker.getRegistration().then(function (r) { r.unregister(); document.location.reload() });
> ```


This method only works if you perform this action with every new start.

**Comment by alexkrass76** [comment]
> I try to create Whatsapp using nativefier 8.0.4, the problem is the same, I used versions of Electron 8.1.1; 9.0.0-beta.9, used whatsapp-fix.js, no result, requires updating chrome


This problem disappeared after running nativefier with the --honest option
Nativefier 8.0.4 Ubuntu 18.04.4 amd64 kernel 5.3.0-050300-generic

**Comment by The-Compiler** [comment]
Note that the above check for the page's text is language-specific. An alternative which works in any language is checking for a link on that page instead:

```js
if (document.querySelector("a[href='https://support.google.com/chrome/answer/95414']"))
```

**Comment by xavery** [comment]
The webapp doesn't work for me regardless of whether I use the extra injected code or not, i.e. it just shows the "update" screen after restarting. The console in developer tools shows this :
```
init.ts:206 Unable to load preload script: /home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js
(anonymous) @ init.ts:206
init.ts:207 Error: Cannot find module 'loglevel'
Require stack:
- /home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js
    at Module._resolveFilename (internal/modules/cjs/loader.js:798)
    at Function../lib/common/reset-search-paths.ts.Module._resolveFilename (reset-search-paths.ts:40)
    at Module._load (internal/modules/cjs/loader.js:691)
    at Module._load (electron/js2c/asar.js:717)
    at Function.Module._load (electron/js2c/asar.js:717)
    at Module.require (internal/modules/cjs/loader.js:853)
    at require (internal/modules/cjs/helpers.js:74)
    at Object.<anonymous> (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:21)
    at Object.<anonymous> (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:68)
    at Module._compile (internal/modules/cjs/loader.js:968)
(anonymous) @ init.ts:207
/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:46 Uncaught ReferenceError: Cannot access 'INJECT_JS_PATH' before initialization
    at injectScripts (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:46)
    at HTMLDocument.<anonymous> (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:16)
```
Could that even be a nativefier issue, or is something else seriously wrong?

**Comment by dospes** [comment]
> 
> 
> The webapp doesn't work for me regardless of whether I use the extra injected code or not, i.e. it just shows the "update" screen after restarting. The console in developer tools shows this :
> 
> ```
> init.ts:206 Unable to load preload script: /home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js
> (anonymous) @ init.ts:206
> init.ts:207 Error: Cannot find module 'loglevel'
> Require stack:
> - /home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js
>     at Module._resolveFilename (internal/modules/cjs/loader.js:798)
>     at Function../lib/common/reset-search-paths.ts.Module._resolveFilename (reset-search-paths.ts:40)
>     at Module._load (internal/modules/cjs/loader.js:691)
>     at Module._load (electron/js2c/asar.js:717)
>     at Function.Module._load (electron/js2c/asar.js:717)
>     at Module.require (internal/modules/cjs/loader.js:853)
>     at require (internal/modules/cjs/helpers.js:74)
>     at Object.<anonymous> (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:21)
>     at Object.<anonymous> (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:68)
>     at Module._compile (internal/modules/cjs/loader.js:968)
> (anonymous) @ init.ts:207
> /home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:46 Uncaught ReferenceError: Cannot access 'INJECT_JS_PATH' before initialization
>     at injectScripts (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:46)
>     at HTMLDocument.<anonymous> (/home/daniel/WhatsAppWeb-linux-x64/resources/app/lib/preload.js:16)
> ```
> 
> Could that even be a nativefier issue, or is something else seriously wrong?

Same problem for me, errors look about the same. The @saulocastelo workaround still works. Anyone have an idea how to fix the injections apparently not loading?

**Comment by raulcraveiro** [comment]
> Note that the above check for the page's text is language-specific. An alternative which works in any language is checking for a link on that page instead:
> 
> ```js
> if (document.querySelector("a[href='https://support.google.com/chrome/answer/95414']"))
> ```

This worked for me! Now, instead of reloading dozens of times, It just reloads one and then It work. <3

**Comment by aarestu** [comment]
```
function recheck () {
        if (document.querySelector("a[href='https://support.google.com/chrome/answer/95414']")) {
                console.log("reload");
                navigator.serviceWorker.getRegistration().then((registration) => {
                        registration.unregister();
                        document.location.reload();
                });
        }
        setTimeout(go, 5000); // callback

}
recheck();
```
this **workaround** worked for me

**Comment by je-vv** [comment]
@aarestu, where to place that code snippet so that it becomes a persistent workaround?

**Comment by aarestu** [comment]
> @aarestu, where to place that code snippet so that it becomes a persistent workaround?

@je-vv just follow @lipe1966 instruction:

> @lipe1966 create file `whatsapp-fix.js` with inject code from @saulocastelo comment, then `nativefier -n "WhatsApp" --inject whatsapp-fix.js "https://web.whatsapp.com"`

**Comment by renyhp** [comment]
Is there a way to inject the code without building the app again?

**Comment by samueleisalinuxuser** [comment]
> Is there a way to inject the code without building the app again?
> 
> EDIT: Yes, there is. Simply insert @aarestu's code in `resources/app/inject/inject.js`.

Ok, thanks a lot, I switched to Telegram years ago, but unfortunately most of my contacts still use whatsapp and that makes me have to use it :( .

**Comment by cr3ative** [comment]
I just had to "get around" this and it's working fine with the "safari" user agent setting. Here's my build command:

```bash
nativefier https://web.whatsapp.com --inject whatsapp.css --title-bar-style "hiddenInset" --name "WhatsApp" --darwin-dark-mode-support --background-color "#36393f" --counter "true" --bounce --user-agent "safari"
```

whatsapp.css:

```css
/* site.css */

/* header is draggable... */
header {
  -webkit-app-region: drag;
}

/* but any buttons inside the header shouldn't be draggable */
header button {
  -webkit-app-region: no-drag;
}

/* perhaps move some items out of way for the traffic light */
#side header > div:first-of-type {
  margin-left: 75px;
}
```

**Comment by nielstron** [comment]
The injection code to make whatsapp work in a nativefier app recently stopped working for me as well.

However, in case anyone else has the same issue, here is my workaround/new injection js (somewhat more reliable then the previous one?):
```js
// unregister serviceworker after startup -> don't interfere at next startup
let deleteFun = function(){
  if("serviceWorker" in navigator){
    navigator.serviceWorker.getRegistrations().then(function(e){
     //console.log(e);
     for(r of e){
       r.unregister();
     }
   });
  }};
deleteFun();
setInterval(deleteFun,1000);
```
(already in https://github.com/nielstron/ubuntu.whatsapp-nativefier)

It might be a good idea to include it in future releases of the package if more people have this issue.

**Comment by Nowaker** [comment]
The problem with Whatsapp is caused by Nativefier, not the injection code we use to make it work. Nativefier fails to correctly load the inject code.

![image](https://user-images.githubusercontent.com/611746/135183446-07eaf686-ba1d-4573-a9b0-adf1815a5f2c.png)

`Uncaught ReferenceError: log is not defined` from https://github.com/nativefier/nativefier/blob/master/app/src/preload.ts#L78

```js
function injectScripts() {
    const needToInject = fs.existsSync(exports.INJECT_DIR);
    if (!needToInject) {
        return;
    }
    // Dynamically require scripts
    try {
        const jsFiles = fs
            .readdirSync(exports.INJECT_DIR, { withFileTypes: true })
            .filter((injectFile) => injectFile.isFile() && injectFile.name.endsWith('.js'))
            .map((jsFileStat) => path.join('..', 'inject', jsFileStat.name));
        for (const jsFile of jsFiles) {
            log.debug('Injecting JS file', jsFile);  // <------------- PROBLEM HERE
            require(jsFile);
        }
    }
    catch (err) {
        log.error('Error encoutered injecting JS files', err);
   }
```

`log` is not defined, so the code never goes to `require(jsFile)`. Instead all you get is:
```
VM1776 preload.js:84 Uncaught ReferenceError: log is not defined
    at injectScripts (VM1776 preload.js:84)
    at HTMLDocument.<anonymous> (VM1776 preload.js:29)
```

I don't know why `log` is undefined - especially given `const log = console;` in the Typescript source - but the nativefied WhatsApp from https://aur.archlinux.org/packages/whatsapp-nativefier is broken for this reason.

For full context, here's the `preload.js` as seen by me:

```js
(function (exports, require, module, __filename, __dirname, process, global, Buffer) { return function (exports, require, module, __filename, __dirname) { "use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.INJECT_DIR = void 0;
/**
 * Preload file that will be executed in the renderer process.
 * Note: This needs to be attached **prior to imports**, as imports
 * would delay the attachment till after the event has been raised.
 */
document.addEventListener('DOMContentLoaded', () => {
    injectScripts(); // eslint-disable-line @typescript-eslint/no-use-before-define
});
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const electron_1 = require("electron");
// Do *NOT* add 3rd-party imports here in preload (except for webpack `externals` like electron).
// They will work during development, but break in the prod build :-/ .
// Electron doc isn't explicit about that, so maybe *we*'re doing something wrong.
// At any rate, that's what we have now. If you want an import here, go ahead, but
// verify that apps built with a non-devbuild nativefier (installed from tarball) work.
// Recipe to monkey around this, assuming you git-cloned nativefier in /opt/nativefier/ :
// cd /opt/nativefier/ && rm -f nativefier-43.1.0.tgz && npm run build && npm pack && mkdir -p ~/n4310/ && cd ~/n4310/ \
//    && rm -rf ./* && npm i /opt/nativefier/nativefier-43.1.0.tgz && ./node_modules/.bin/nativefier 'google.com'
// See https://github.com/nativefier/nativefier/issues/1175
// and https://www.electronjs.org/docs/api/browser-window#new-browserwindowoptions / preload
exports.INJECT_DIR = path.join(__dirname, '..', 'inject');
/**
 * Patches window.Notification to:
 * - set a callback on a new Notification
 * - set a callback for clicks on notifications
 * @param createCallback
 * @param clickCallback
 */
function setNotificationCallback(createCallback, clickCallback) {
    const OldNotify = window.Notification;
    const newNotify = function (title, opt) {
        createCallback(title, opt);
        const instance = new OldNotify(title, opt);
        instance.addEventListener('click', clickCallback);
        return instance;
    };
    newNotify.requestPermission = OldNotify.requestPermission.bind(OldNotify);
    Object.defineProperty(newNotify, 'permission', {
        get: () => OldNotify.permission,
    });
    // @ts-expect-error TypeScript says its not compatible, but it works?
    window.Notification = newNotify;
}
function injectScripts() {
    const needToInject = fs.existsSync(exports.INJECT_DIR);
    if (!needToInject) {
        return;
    }
    // Dynamically require scripts
    try {
        const jsFiles = fs
            .readdirSync(exports.INJECT_DIR, { withFileTypes: true })
            .filter((injectFile) => injectFile.isFile() && injectFile.name.endsWith('.js'))
            .map((jsFileStat) => path.join('..', 'inject', jsFileStat.name));
        for (const jsFile of jsFiles) {
            log.debug('Injecting JS file', jsFile);
            require(jsFile);
        }
    }
    catch (err) {
        log.error('Error encoutered injecting JS files', err);
    }
}
function notifyNotificationCreate(title, opt) {
    electron_1.ipcRenderer.send('notification', title, opt);
}
function notifyNotificationClick() {
    electron_1.ipcRenderer.send('notification-click');
}
// @ts-expect-error TypeScript thinks these are incompatible but they aren't
setNotificationCallback(notifyNotificationCreate, notifyNotificationClick);
electron_1.ipcRenderer.on('params', (event, message) => {
    log.debug('ipcRenderer.params', { event, message });
    const appArgs = JSON.parse(message);
    log.info('nativefier.json', appArgs);
});
electron_1.ipcRenderer.on('debug', (event, message) => {
    log.debug('ipcRenderer.debug', { event, message });
});
//# sourceMappingURL=preload.js.map
}.call(this, exports, require, module, __filename, __dirname); });
```

**Comment by TheCleric** [comment]
> The problem with Whatsapp is caused by Nativefier, not the injection code we use to make it work. Nativefier fails to correctly load the inject code.
> I don't know why `log` is undefined - especially given `const log = console;` in the Typescript source - but the nativefied WhatsApp from https://aur.archlinux.org/packages/whatsapp-nativefier is broken for this reason.

@Nowaker, I agree with your assessment. Not sure what is causing that log definition to be missing as we added that 4 months ago (and looking through the Arch deps, they should be on a version that includes it). We could add a workaround, but my fear would be that somehow that code wouldn't make it upstream just as the `const log = console;` does not.

**Comment by Nowaker** [comment]
Let's find out why `log` is undefined. [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nodejs-nativefier) for `nodejs-nativefier` doesn't do anything crazy. It purely uses your project, without modifications:

```
source=(http://registry.npmjs.org/$_npmname/-/$_npmname-$_npmver.tgz)
npm install -g --cache "${srcdir}/npm-cache" --prefix "$pkgdir/usr" "$srcdir/$_npmname-$pkgver.tgz"
```

I've just built the latest version, 45.0.4, just to make sure.
```diff
-_npmver=45.0.1
+_npmver=45.0.4
-sha256sums=('7e2faca780c66fa6188bd7e55093822cffcee950c008240c21799ab5edf312e6')
+sha256sums=('SKIP')
```
But still no change. `log` is still undefined: 
![image](https://user-images.githubusercontent.com/611746/135328649-5d47ef90-b188-4938-bf91-94b8cb8400b9.png)

```
% cat /opt/whatsapp-nativefier/resources/app/lib/preload.js|grep log
            log.debug('Injecting JS file', jsFile);
        log.error('Error encoutered injecting JS files', err);
    log.debug('ipcRenderer.params', { event, message });
    log.info('nativefier.json', appArgs);
    log.debug('ipcRenderer.debug', { event, message });
```

**Comment by Nowaker** [comment]
Okay... I've spotted something silly in Whatsapp's [PKGBUILD](https://github.com/frealgagu/archlinux.whatsapp-nativefier/blob/master/PKGBUILD#L47).

```diff
  sed -i -e "/loglevel/d" "${srcdir}/${_folder}/resources/app/lib/preload.js"
```

This removes the line that is very required... Checking it now.

**Comment by Nowaker** [comment]
Yeah, that fixes the problem. Not sure why it was there. Submitting a patch to the PKGBUILD.

**Comment by AI-ien** [comment]
For me only code snippet from [nielstron comment](https://github.com/nativefier/nativefier/issues/719#issuecomment-894724198) worked using safari user agent. Anything else just fails, also code snippet on main README.md page.
